README.md

```diff
@@ -1,2 +1,58 @@
+
+This repository contains information and a Python script for designing primers to generate mutant libraries of genes with specific desired mutations at each site.
+
+## Codon mutagenesis
+
+Codon mutagenesis can be used to create mutant libraries of a gene with all possible codon mutations.
+Such libraries are useful for experiments such as [deep mutational scanning](https://www.ncbi.nlm.nih.gov/pubmed/25075907) or [mutational antigenic profiling](http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271)
+
+A [script in a separate repository](https://github.com/jbloomlab/CodonTilingPrimers) can be used to generate random mutations at each codon in a gene. This protocol randomly mutates each codon to all other possible codons. This protocol and papers using it are also described in that repository.
+
+Rather than generating mutant libraries with all possible codon mutations at every site in a gene, the script in this repository can be used to generate primers that will result in specific desired mutations at each site in a gene. In this way, a mutant library can be generated which has fewer lethal mutations than random codon mutagenesis.
+
+## Running the script to design primers
+
+The [create_primers.py](create_primers.py) Python script can be used to create primers that tile the codons of a gene in both the forward and reverse direction. Each pair of foward and reverse primers will produce one specified mutation at one specified site. The desired mutations must be given as a csv file argument.
+
+### Required arguments
+
+*seq* : Text file containing the sequence of the gene. The gene itself should be upper case, including the start and stop codons. Any flanking regions should be lower case. There must be >= (primerlength - 3) / 2.0 nucleotides before the first codon mutagenized and after the last codon mutagenized, otherwise there would not be enough flanking nucleotides for a primer to mutate those codons.
+
+*mutations_csv* : csv file containing a table with the mutations the primers will make to seq. The table should have a column, 'site', with site numbers, and a column, 'mutant', with the single letter amino acid mutant to make at that site. Each site can have multiple mutants, or none. The site number should denote the number of the codon in the uppercase gene, with the start codon being 1.
+
+*codon_frequency_csv* : csv file containing a table with codon frequencies used to determine which codons sites will be mutated to. The table should have the columns 'aa', 'codon', and 'frequency'. Each row should have one single letter amino acid in 'aa', one codon corresponding to that amino acid in 'codon', and the frequency of that codon in 'frequency'. The script uses the codon with the highest frequency to replace the codon at a site to make mutations. Codon frequency tables for different organisms can be found [here](https://www.kazusa.or.jp/codon/).
```

Comment:
I think BioPython can be used to generate codon frequency tables or use them to score sequences on their codon usage, but I did not find any well documented functionality to back translate using them. If you can find something like that I would use it.

Reply:
I need to think more about the back translation, but why exactly is it necessary? Aren't you specifying mutants by aa, so why do you need to back translate from codon?

Reply:
If you specify what mutants you want by aa, then unless you want to make primers for each codon for that aa, you have to pick a codon (back translate the aa to a codon). The codon frequency table is used so that the codon picked isn't just random, but is the most frequent codon for that aa in whatever organism you supply the table for. I am planning on adding an option for just random, or semi random like you suggested.

Reply:
Gotcha. I thought that we had used a built in function for this in `serologyseq` (https://github.com/jbloomlab/serologyseq), but it looks like we did write our own dictionary to map amino acids and codons. We did just use built in cai tables though, so there might be something helpful in the `design_oligos.py` script (https://github.com/jbloomlab/serologyseq/blob/master/src/design_oligos.py) if you didn't want to have to have the user input their own codon_frequency table, but it's not as easy as I thought it was, so what you have with the user inputting their own frequency table is also fine.